redis/src/aio/connection.rs

```diff
@@ -256,108 +267,205 @@ where
 }
 
 /// Represents a `PubSub` connection.
-pub struct PubSub<C = Pin<Box<dyn AsyncStream + Send + Sync>>>(Connection<C>);
-
-/// Represents a `Monitor` connection.
-pub struct Monitor<C = Pin<Box<dyn AsyncStream + Send + Sync>>>(Connection<C>);
+pub struct PubSub<C = Pin<Box<dyn AsyncStream + Send + Sync>>> {
+    framed: Framed<C, ValueCodec>,
+    queue: VecDeque<Msg>,
+}
 
 impl<C> PubSub<C>
 where
     C: Unpin + AsyncRead + AsyncWrite + Send,
 {
-    fn new(con: Connection<C>) -> Self {
-        Self(con)
+    fn new(con: Connection<C>) -> PubSub<C> {
+        let framed = ValueCodec::default().framed(con.con);
+
+        PubSub {
+            framed,
+            queue: Default::default(),
+        }
+    }
+
+    // Deliver the PUBSUB command to this PUBSUB connection
+    async fn command<'a>(&'a mut self, cmd: &'a Cmd) -> RedisResult<()> {
+        let mut out = Vec::new();
+
+        cmd.write_packed_command(&mut out);
+        self.framed.send(out).await?;
+
+        self.next_response().await
+    }
+
+    async fn next_response(&mut self) -> RedisResult<()> {
+        while let Some(item) = self.framed.next().await {
+            match item {
+                Ok(Ok(value)) => match Msg::from_value(&value) {
+                    Some(msg) => self.queue.push_front(msg),
+                    _ => return Ok(()),
```

Comment:
please add a todo to verify this isn't Value::Error after https://github.com/redis-rs/redis-rs/pull/813 will be merged